subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java

```diff
@@ -103,86 +129,79 @@ private void resizeTo(int newBuildProgressLabelCount) {
 
+    private void attach(ProgressOperation operation, StyledLabel label) {
+        AssociationLabel association = new AssociationLabel(operation, label);
+        operationIdToAssignedLabels.put(operation.getOperationId(), association);
+    }
+
     private void detach(ProgressOperation operation) {
         if (operation.getParent() != null) {
-            removeDirectChildOperationId(operation.getParent().getOperationId(), operation.getOperationId());
+            registry.find(operation.getParent().getOperationId()).remove(operation.getOperationId());
         }
 
-        AssociationLabel association = operationIdToAssignedLabels.remove(operation.getOperationId());
-        if (association != null) {
-            unusedProgressLabels.push(association.label);
-            if (operation.getParent() != null) {
-                attach(operation.getParent());
-            } else if (!unassignedProgressOperations.isEmpty()){
-                attach(unassignedProgressOperations.pop());
-            }
-        } else {
-            unassignedProgressOperations.remove(operation);
+        if (!isRenderable(operation)) {
+            return;
         }
-    }
 
-    private void addDirectChildOperationId(OperationIdentifier parentId, OperationIdentifier childId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children == null) {
-            children = new HashSet<OperationIdentifier>();
-            parentIdToChildrenIds.put(parentId, children);
+        detach(operation.getOperationId());
+        unassignedProgressOperations.remove(operation);
+
+        if (operation.getParent() != null) {
+            attach(operation.getParent());
+        } else if (!unassignedProgressOperations.isEmpty()) {
+            attach(unassignedProgressOperations.pop());
         }
-        children.add(childId);
     }
 
-    private void removeDirectChildOperationId(OperationIdentifier parentId, OperationIdentifier childId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children == null) {
-            return;
-        }
-        children.remove(childId);
-        if (children.isEmpty()) {
-            parentIdToChildrenIds.remove(parentId);
+    private void detach(OperationIdentifier operationId) {
+        AssociationLabel association = operationIdToAssignedLabels.remove(operationId);
+        if (association != null) {
+            unusedProgressLabels.push(association.label);
         }
     }
 
-    private boolean isChildAssociationAlreadyExists(OperationIdentifier parentId) {
-        Set<OperationIdentifier> children = parentIdToChildrenIds.get(parentId);
-        if (children != null && !children.isEmpty()) {
-            return true;
+    // Any ProgressOperation in the parent chain has a message, the operation is considered renderable.
+    private boolean isRenderable(ProgressOperation operation) {
+        for (ProgressOperation current = operation;
```

Comment:
I'm unsure how this can be unified with `BuildStatusRenderer`. Could you help me see through this?